### PR TITLE
WT-13031 Create dist script to check sub-system versus across directory naming

### PR DESCRIPTION
**Pull request by pmacko86**
This PR is an extension of WT-12986, which made existing, ad hoc module boundaries and APIs more visible by renaming functions that are currently used only within their respective modules to use prefix `__wti` instead of `__wt` (just capturing the current state instead of trying to define what it should be). This PR adds a check to `s_all` that this renaming is all correct.

The script checks that:
* Functions beginning with `__wti` are only used or called by other files within that directory.
* Functions beginning with `__wti` are used in at least two files (otherwise they could be made static).
* Functions beginning with `__wt` are used outside of their directory (otherwise they could be made `__wti`).

It uses the same exceptions list as `s_funcs` (i.e., `s_funcs.list`), which ensures that all `extern` functions are used.

### CODE REVIEW

File: dist/function.py
Comment (sueloverso) [inline dist/function.py:203]: It took a lot of attention to follow exactly what this is doing. It looks great. I would appreciate a bunch more comments throughout this function.

**Review by sueloverso — APPROVED** [review]
LGTM thanks for the quick turnaround and fixing the ones I missed. One small comment.

File: dist/function.py
Comment (pmacko86) [reply dist/function.py:203]: Sounds good; I added more comments in [7f37108](https://github.com/wiredtiger/wiredtiger/pull/10667/commits/7f37108e0ab8c074783c4ceae7b4c7efcbe41532). Hopefully they are indeed helpful!

File: dist/function.py
Comment (ershov) [inline dist/function.py:312]: Please don't enforce the reduction of scope.

**Review by ershov — CHANGES_REQUESTED** [review]
Thanks Peter for the PR, it's a big shift towards a better codebase!

Unfortunately there's no summary of what and how it does, so I try to read it from the code and comments.

It appears that your PR not only checks the correct usage of scoped functions but also enforces setting the scope to the currently detected minimum.

I have concerns about this type of enforcement because it relates to the module's interface design rather than how the functions are currently being used.

For example, I might create a small helper function that I don't want to be called externally, or an interface function intended for broader use, even if it's not currently used elsewhere.

I'm fine with enforcing the former, but I oppose automation forcing me to limit my interface functions to local scope based solely on their current usage.

**Comment by sueloverso** [comment]
@ershov if you have something that is for some future use, why can't it be named as its current internal function and then renamed if/when the future work is done that calls it outside the module? I think not enforcing things makes the script fairly meaningless then if it is just ignored. Can you give me an example to help understand what you mean?

**Comment by ershov** [comment]
@sueloverso Sure. Suppose I'm writing a module that sorts numbers. The implementation looks like this:

```c
/*
 * Function that does X
 * @private
 */
void __sort_helper_X(...) {
  ...
}

/*
 * Function to compare numbers
 * @private
 */
void __sort_compare_fn(...) {
  ...
  __sort_helper_X(...)
  ...
}

/*
 * Function to sort numbers
 * @private
 */
void __sort_internal(int *begin, int *end) {
  ...
  __sort_compare_fn(...)
  ...
}

/*
 * Function to sort numbers
 * @public
 */
void sort(int *array, size_t length) {
  ...
  __sort_internal(...)
  ...
}
```

This is a well designed module with its external interface and internal helpers. It's easy to read and understand. The purpose and the scope of functions is clear and it comes **from design**, not from the external use.

I don't really see why if an interface function is unused at the moment, it must become an internal one whereas it's designed to be externally visible.

If one needs to rename functions based on external conditions, it's not a well designed code and it's not well maintainable.

I see the purpose of this checker as protecting internal (private) functions from being used from outside the scope but not forcing to reduce the scope.

**Comment by ershov** [comment]
Another example is containers. A typical container has interface functions to add, remove, access, sort, search elements. If the existing code doesn't use some of these functions, the interface is still the interface and internal functions are internal.

**Comment by pmacko86** [comment]
I appreciate the examples, as they make it more concrete!

This is where I've used the existing ignore list (s_funcs.list) in the past to get around the existing check that ensures that all functions are used somewhere. The list is also useful for functions that are used for debugging purposes.

If the ignore list is insufficient, we can ignore entire directories from the check, such as `src/support`, which is most likely to contain the kind of functions that you described. And then continue to use the ignore list for the rest.

I'd like to keep the check, as I agree with Sue that without it people may just end up introducing `__wt_` functions unnecessarily.

**Comment by ershov** [comment]
@pmacko86 Adding all interface functions to the global ignore list could be ok as a temporary measure but I'm not sure it's a good way to express the module's design intents. There are better options like reading that from the function's comment. If someone later decides to change the function's scope, it's all in one place.

Another big concern is that this PR doesn't consider whether a particular function is actually designed to be externally visible (or how much externally) and should be added to the ignore list or its scope should be reduced. This PR just reduces the scope for all in bulk. This is going to introduce confusion in future where it's done incorrectly.

I understand that considering each function individually is time consuming. Assuming that the aim of this change is to improve the code quality, we shouldn't be rushed and actually degrade it. A safer alternative could be bulk-adding all these "incorrectly scoped" functions to the ignore list, then go one by one and change the scope where needed.

Speaking about "introducing __wt_ functions unnecessarily", I prefer to go the other way around: "allow the developers to limit the scope where they explicitly want to" rather than "forbid everything and pinpoint externally available stuff". This might be a matter of personal preference and is open for discussion.

**Comment by korteland** [comment]
I just caught up on reading this and I do agree with Yury's points.

My understanding is that one part of the goal is to stop people accidentally making functions "too public" - either by using a `__wt_` prefix instead of `__wti_`, or by having it non-static when it's not used outside a file. I don't think these goals are wrong, but the rigid enforcement bothers me slightly - it is harder to do a longer, multi-PR change (e.g. adding an interface in one ticket, using it in another). I think it also adds friction when it becomes necessary to expose something. For example, consider `__wt_raw_to_nhex`. It's a function that is used outside the `support/` module, and if `__wt_nhex_to_raw` was ever needed it would make sense to give it the same visibility even if nothing outside of `support/` needed it. But it doesn't seem that would be possible with this change?

I'd normally be all for codifying enforcements of coding standards, even when the standard is flexible (see all the ignore/exception lists in `dist/`). But this script doesn't seem to come with that "exception list" escape hatch, and going back/forth when a function becomes used or unused seems like a noisy process since all callsites must also be modified. 

Secondly, this looks like a "bulk change" in that it doesn't seem much in-depth review has been done - the script has been applied, and the necessary changes made to stop it complaining. That's not bad as a starting point but I think a more thoughtful, per-function review would be more effective.

Finally, there's another team also working on reviewing the modularity situation, and while I don't think this work conflicts, I'd prefer to give that team the space to find things and make recommendations on their own - perhaps there's a better way to go about this. Is there a rush on this PR? Normally a change in coding standards like this would get socialised in the team before a PR gets dropped.

File: dist/function.py
Comment (ddanderson) [inline dist/function.py:309]: I think the comment is opposite what it should be:  `functions that _are_ allowed to be used anywhere`?

**Review by ddanderson — APPROVED** [review]
LGTM.  Very readable Python.

**Comment by sueloverso** [comment]
@pmacko86 perhaps this work can be split into two tickets. This ticket remains for the script and a separate ticket/PR to fix the inconsistencies it found. Those inconsistency fixes should be fixed ASAP and only the script details can then be the focus here.

File: dist/function.py
Comment (tod-johnson-mongodb) [inline dist/function.py:304]: Change and module != fn.module to and fn.fn.used_outside_of_module.

**Comment by sueloverso** [comment]
My general feeling is that, as @pmacko86 mentioned, we have a mechanism for dealing with future-looking functions with the list of functions to ignore, and his script takes that into account. So any of these theoretical multi-PR reasons would already have to use that in order to quiet things. This script does not make it any worse/better but wisely uses the existing mechanism. There are many parts of `s_all` that enforce various standards, and then have ways around them in the exceptional cases. Again, this script does not change any of that because it uses the exact same mechanism for the exceptions. It would be much worse to introduce another mechanism. 

I don't understand or see any connection between this one script and the modularity work that Jeremy's team is doing. What connection do you see?

Having the script not enforce something means its output can be ignored and if that ignored warning is then merged to develop everyone on the team  is subject to the "ignored warning". How do you see the script working and being effective in that situation? What should the script do when it finds something that is inconsistent that then gets merged?

File: dist/function.py
Comment (tod-johnson-mongodb) [inline dist/function.py:199]: A possible performance improvement.
I wonder if there should be two more fields: self.global_visible for __wt names and self.module_visible for __wti names. Then check these instead of using startswith(). Which is faster?

File: dist/function.py
Comment (tod-johnson-mongodb) [inline dist/function.py:243]: These lines could be omitted since only functions that are start with __wt or __wti are in fn_defs. Would that be faster? How does two starswith() compare with one in dict()? If you delete these lines add a comment that functions that are not __wt and __wti are filtered out by in fnd_defs.

File: dist/function.py
Comment (tod-johnson-mongodb) [inline dist/function.py:305]: This message can be printed multiple times for the same file. Either fix that or also print the line number.

File: dist/function.py
Comment (tod-johnson-mongodb) [inline dist/function.py:309]: How about:
# Load the list of functions that can have scope larger than their actual usage.
or
# Load the list of functions whose scope is not enforced.

File: dist/function.py
Comment (tod-johnson-mongodb) [reply dist/function.py:243]: I was wrong. These lines are needed for line 249. They could be changed to the below and left here of moved just before lien 249.

   global_scope = fn_name.startswith('__wt_')
   if not global_scope and not fn_name.startswith('__wti_'):
        continue

Then change line 249
     fn_defs[fn_name] = FunctionDefinition(fn_name, module, source_file, global_scope)

Instead of checking fn_name.startswith('__wt__') check fn_defs[fn_name].global_scope.
Instead of checking fn_name.startswith('__wti__') check not fn_defs[fn_name].global_scope.

**Review by tod-johnson-mongodb — APPROVED** [review]
I made some suggestions. They are all optional. I think the most important one is the one changing # Load the list of functions that are not allowed to be used anywhere.

**Comment by sueloverso** [comment]
@ershov and @korteland I have been re-reading the discussion here for the reason to block this PR. I need help understanding it. You both mention future use functions and multi-PR merges. That is fine and that happens. But today, if you have public interface `wt_future_function` that is written but not called anywhere, and `wti_future_intrause` that is written but not called anywhere, the `s_funcs` script will prevent you from merging. You will need to add them to `s_funcs.list` and then you can merge. That is the state of the world today on `develop` and has been for forever.

With Peter's script merged, NOTHING changes in this situation. The `s_funcs` script will still require you to add them to `s_funcs.list` and the new script will not send a warning about them nor prevent merging. If someone really adds a real function that is called and used then it will be checked. 

Please help me understand what you believe is different with the existence of this script in the 'future function' case?

**Comment by pmacko86** [comment]
> @pmacko86 perhaps this work can be split into two tickets. This ticket remains for the script and a separate ticket/PR to fix the inconsistencies it found. Those inconsistency fixes should be fixed ASAP and only the script details can then be the focus here.

Per @sueloverso's suggestion, I created [WT-13088](https://jira.mongodb.org/browse/WT-13088) for just the inconsistencies given the desire to get these fixes in first, so that this PR can focus entirely on the script. (I still have to remove those changes from this PR.)

**Comment by ershov** [comment]
Ok now that I talked to multiple people, I think that I was able to derive the actual intent of this change:

The actual intent of this series of changes (WT-6722, WT-12986, WT-13031) is **not** the modularisation of source code via assigning the functions their public/protected/private visibility scope but rather manifesting the current scope of calls to the functions in their prefixes.

If this is correct, I withdraw all my comments.

I still think that the PR description should not be empty, please put a description of what it does and how it does it.

**Comment by korteland** [comment]
> I don't understand or see any connection between this one script and the modularity work that Jeremy's team is doing. What connection do you see?

No direct connection. My only objection here (and it's minor) is that it'd be preferable to do the exploration first before getting PRs in to address issues. But now that I realise it's an extension of previous work, I think we can disregard this point.

I'm in much the same boat as Yury in that was very hard to figure out the intent of this change from the PR description (empty) and the Jira tickets (which had descriptions, but were fairly terse). I initially misunderstood it. I also didn't realise it reused the existing exception list, I was looking for a new list in the diff. Please don't let my earlier comments block merging this.

**Comment by pmacko86** [comment]
Thank you all!

Yes, the purpose of this PR is to add a check that the renaming done in [WT-12986](https://jira.mongodb.org/browse/WT-12986) is indeed correct, with the goal of making existing, ad hoc module boundaries and APIs more visible—to capture the current state, not to define any future modularity.

I added a PR comment as requested, and I will add something like that as a comment to {{function.py}} as well, including a note about the exceptions list.

@ershov, @korteland, please let me know if there is anything else that I should add to the PR comment and/or to the script itself. I'll update the script with these comments and work on the other requested changes early next week (I don't think I'd be able to get to it before the weekend).

**Review by ershov — APPROVED** [review]
Thanks Peter!
Your description to this PR makes it perfectly clear! This is really Looking Good To Me :)

Side note: This is not a part of this PR but it'd be also great to update [WT-12986](https://jira.mongodb.org/browse/WT-12986) so that it's also as easy to understand.
It'd be great help for future ourselves looking back to when these changes were made.

**Review by korteland — APPROVED** [review]
LGTM - sorry for the noise when I misunderstood the purpose.

File: dist/function.py
Comment (pmacko86) [reply dist/function.py]: Fixed in [acfc825](https://github.com/wiredtiger/wiredtiger/pull/10667/commits/acfc825f968f01aa6feb302fe1f275bfe33ad7f3).

File: dist/function.py
Comment (pmacko86) [reply dist/function.py]: Fixed in [acfc825](https://github.com/wiredtiger/wiredtiger/pull/10667/commits/acfc825f968f01aa6feb302fe1f275bfe33ad7f3). I used a different suggestion proposed by Tod.

File: dist/function.py
Comment (pmacko86) [reply dist/function.py]: Fixed in [acfc825](https://github.com/wiredtiger/wiredtiger/pull/10667/commits/acfc825f968f01aa6feb302fe1f275bfe33ad7f3).

File: dist/function.py
Comment (pmacko86) [reply dist/function.py:199]: Fixed in [acfc825](https://github.com/wiredtiger/wiredtiger/pull/10667/commits/acfc825f968f01aa6feb302fe1f275bfe33ad7f3).

File: dist/function.py
Comment (pmacko86) [reply dist/function.py]: Fixed in [acfc825](https://github.com/wiredtiger/wiredtiger/pull/10667/commits/acfc825f968f01aa6feb302fe1f275bfe33ad7f3). I added a line number.

File: dist/function.py
Comment (pmacko86) [reply dist/function.py]: Fixed in [acfc825](https://github.com/wiredtiger/wiredtiger/pull/10667/commits/acfc825f968f01aa6feb302fe1f275bfe33ad7f3). That's a much better way to put it!

File: dist/dist.py
Comment (pmacko86) [reply dist/dist.py:35]: That would be nice! `src/btree` is big, so it would be nice to break it up this way. After that, we can modify function `infer_module` in `function.py` to account for this.

**Review by ddanderson — APPROVED** [review]
LGTM!

**Pull request merged**
